### PR TITLE
Iox #835 json config reader

### PR DESCRIPTION
**Pull request by oschuene**
## Pre-Review Checklist for the PR Author

1. [ ] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [ ] Tests follow the [best practice for testing][testing]
1. [ ] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [ ] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [ ] Commits messages are according to this [guideline][commit-guidelines]
    - [ ] Commit messages have the issue ID (`iox-#123 commit text`)
    - [ ] Commit messages are signed (`git commit -s`)
    - [ ] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [ ] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [ ] Relevant issues are linked
1. [ ] Add sensible notes for the reviewer
1. [ ] All checks have passed (except `task-list-completed`)
1. [ ] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CHANGELOG.md

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Closes **TBD**

### CODE REVIEW

**Comment by dkroenke** [comment]
@oschuene Thanks for the Contribution!
Only minor remarks on the first sight:
- Please link this PR to the corresponding GitHub issue (https://github.com/eclipse-iceoryx/iceoryx/issues/835) by editing the PR and replacing `Closes TBD` with the ticket number
- You can already fill out the checkboxes in the Pre-Review Checklist for the PR Author

File: iceoryx_posh/test/moduletests/test_roudi_config_json_file_provider.cpp [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+        tmpFile.close();
+
+        auto cmdLineArgs = cmdLineParser.parse(argc, argv);
+
+        JsonRouDiConfigFileProvider reader(cmdLineArgs.value());
+        return reader.parse();
+    }
+};
+
+TEST_F(JsonConfig_test, TestReader)
```
Comment (elfenpiff) [inline iceoryx_posh/test/moduletests/test_roudi_config_json_file_provider.cpp:117]: I think there are some tests missing. Please read our guide: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

And then add tests with a mempool config of zero, one, many, limit (MAX_NUMBERS_OF_MEMPOOLS) and limit + 1.

I think the same goes for segment.


Additionally we require tests for entries with a negative count or size of mempools or where the count/size exceeds the storage capacity of an `uint32_t`.

**Review by elfenpiff — CHANGES_REQUESTED** [review]

File: iceoryx_posh/CMakeLists.txt [hunk 2]
```diff
@@ -383,6 +436,10 @@ if(TOML_CONFIG)
     set(ROUDI_EXPORT iceoryx_posh_config iox-roudi)
 endif()
 
+if (JSON_CONFIG)
+    set(ROUDI_EXPORT ${ROUDI_EXPORT} iceoryx_posh_config_json iox-roudi-json)
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/CMakeLists.txt:440]: @elfenpiff unfortunately nobody addressed my concerns from here https://github.com/eclipse-iceoryx/iceoryx/issues/835#issuecomment-855870578

I don't think it is a good idea to have both in parallel. Having multiple ways to do the same thing confuses the user. We should decide on one default config file type and advertise that. If someone wants to use a different one, that should be a cmake option. But I feel especially with the config file we don't do us a favor if we have multiple config file types. How many project you know do that and don't use a 3rd party abstraction, e.g. `QSettings`?

Instead of having a `JSON_CONFIG` and a `TOML_CONFIG` I would rather have a `CONFIG_FILE` option which could either be set to `TOML`, `JSON` or `NONE`. This makes it easier to extend in the future in case we decide to do so. With this, there would be only one `iox-roudi`, independent from what is used as config file.

We just have to decide what we want to have as default. I would vote for `TOML` (config file format) since it has some advantages compared to `JSON` (data exchange format), e.g. it supports comments. See also [first result for `config file formats` on google](https://jhall.io/posts/best-config-file-formats/). If there are issues with `cpptoml`, there are other options we can switch to e.g. `tomlc99` which is validated against two big test suites.

Additionally, we have to decide what to do with the non-default config file type if we extend the options in the default config file type. Do we have to also add all the options to the non-default one? Would this be a release blocker? If yes, how about when we get a third config file upstreamed, e.g. `lua` and a fourth one, e.g. `yaml` and a fifth one, e.g. `inl` and a sixth one, e.g. `XML`?

The same also applies to our gateways. At the moment there is only the Cyclone DDS gateway but in case the Fast DDS gateway gets merged one day we have to ask the same questions.

File: iceoryx_posh/source/roudi/application/roudi_main_json.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    using iox::roudi::IceOryxRouDiApp;
+
+    iox::config::CmdLineParserConfigFileOption cmdLineParser;
+    auto cmdLineArgs = cmdLineParser.parse(argc, argv);
+    if (cmdLineArgs.has_error() && (cmdLineArgs.get_error() != iox::config::CmdLineParserResult::INFO_OUTPUT_ONLY))
+    {
+        iox::LogFatal() << "Unable to parse command line arguments!";
+        return EXIT_FAILURE;
+    }
+    iox::roudi::JsonRouDiConfigFileProvider configFileProvider(cmdLineArgs.value());
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/source/roudi/application/roudi_main_json.cpp:36]: @elfenpiff bad idea. If no file is provided by cmd line args, it is looked if `/etc/iceoryx/roudi_config.toml` is available. Let's assume we also have a `/etc/iceoryx/roudi_config.json`, which one do we choose? Coin flip? Alphabetic order? Well, what if one system has one and another one has accidentally both files? You can just hope that the guy who is debugging this isn't a psychopath and knows where you are living ;)

Fun fact. I once had a dependency to a library with only one function which not everyone was allowed to get. The solution was to create a shim library with the function returning `0xAFFE` (German for ape/monkey/idiot). The user just had to replace that shim library with the real one. Guess what happened. At some day I got a mail from a guy who always got `0xAFFE` on one of the systems and after one week of debugging he asked for help. For some weird reason the shim library was also at that system and found before the real one and therefore also used. Luckily he took it with humor :smile:

**Comment by elBoberido** [comment]
I didn't do a review, just looked it there is a `COLCON_IGNORE`. There isn't. Please add one else there will be issues with ROS.

File: iceoryx_posh/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_posh/CMakeLists.txt:440]: 1. I support the idea with the cmake option `CONFIG_FILE` with the 3 values `TOML`, `JSON` and `NONE`.
2. I also vote for `TOML` as default option.

How should we proceed here with the consistent config options? For this we definitely require some kind of abstraction where maybe a unified abstract path for each setting is described like `general.mempool.chunk.count : uint32_t` (maybe as vector) which is then translated into a TOML or JSON entry by the actual config parser.
In my opinion it is important that we describe all the config paths/entries only once with their type and translate them to toml/json otherwise we will always have consistency issues since some entries are slightly different named or some forget to add an entry or miss-spelled it.
@oschuene @elBoberido what do you think about this approach? @oschuene do you have time to add such an abstraction?

[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/CMakeLists.txt:440]: @elfenpiff @oschuene on a second thought, maybe we still need `JSON_CONFIG` and `TOML_CONFIG` in order to build both at the same time in the CI and run the tests, else we would need to introduce another job just for this. But I would still make it a cmake option to decide which config file type to use for `iox-roudi`.

Regarding the abstraction. I'm not sure I got what you mean :smile:

**Comment by oschuene** [comment]
This pull-request became obsolete. I will close it. Thanks for the many comments.
